# Data cleaning - missing and noisy data
- Missing data are observations which are absent from the dataset.
- These mainly arise due to faulty data collection strategies or the unavailability of values.
- Noise is a general term encompassing outliers, distorted values, inconsistent units of measurement.

Demonstration: Using the open-sourced titanic dataset to demonstrate various data cleaning strategies. The steps that are (usually) followed are:
1. Loading the dataset
2. Getting an overview of the dataset
3. Separate the target variable from the remaining columns
4. Finding the utility of the columns, if needed
5. Extracting high level information about the dataset
6. Check for duplicate records
7. Handling missing values
8. Handling invalid values
9. Handling outliers
10. Encoding categorical variables
11. Removing irrelevant columns
12. Bining and smoothing

## Loading the dataset

In [145]:
path_of_dataset = "titanic1.csv"

In [146]:
import pandas as pd
import numpy as np

In [147]:
df_titanic = pd.read_csv(path_of_dataset)

## Getting an overview of the dataset

In [148]:
df_titanic.shape

(891, 12)

*There are 891 samples and 14 attributes in this dataset.*

In [149]:
df_titanic.head(8)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S

In [150]:
df_titanic.tail(6)

PassengerId  Survived  Pclass                                      Name  \
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket    Fare Cabin Embarked  
885  female  39.0      0      5      382652  29.125   NaN        Q  
886    male  27.0      0      0      211536  13.000   NaN        S  
887  female  19.0      0      0      112053  30.000   B42        S  
888  female   NaN      1      2  W./C. 6607  23.450   NaN        S  
889    male  26.0      0      0      111369  30.000  C148        C  
890    male  32.0      0      0      370376   7.750   NaN        Q

In [151]:
df_titanic.sample(10)

PassengerId  Survived  Pclass  \
252          253         0       1   
399          400         1       2   
235          236         0       3   
580          581         1       2   
583          584         0       1   
276          277         0       3   
571          572         1       1   
599          600         1       1   
768          769         0       3   
135          136         0       2   

                                              Name     Sex   Age  SibSp  \
252                      Stead, Mr. William Thomas    male  62.0      0   
399               Trout, Mrs. William H (Jessie L)  female  28.0      0   
235                   Harknett, Miss. Alice Phoebe  female   NaN      0   
580                    Christy, Miss. Julie Rachel  female  25.0      1   
583                            Ross, Mr. John Hugo    male  36.0      0   
276              Lindblom, Miss. Augusta Charlotta  female  45.0      0   
571  Appleton, Mrs. Edward Dale (Charlotte Lamson)  female  53.0      2   
599   Duff Gordon, Sir. Cosmo Edmund ("Mr Morgan")    male  49.0      1   
768                            Moran, Mr. Daniel J    male   NaN      1   
135                             Richard, Mr. Emile    male  23.0      0   

     Parch         Ticket     Fare Cabin Embarked  
252      0         113514  26.5500   C87        S  
399      0         240929  12.6500   NaN        S  
235      0     W./C. 6609   7.5500   NaN        S  
580      1         237789  30.0000   NaN        S  
583      0          13049  40.1250   A10        C  
276      0         347073   7.7500   NaN        S  
571      0          11769  51.4792  C101        S  
599      0       PC 17485  56.9292   A20        C  
768      0         371110  24.1500   NaN        Q  
135      0  SC/PARIS 2133  15.0458   NaN        C

<i>The dataset contains missing values, denoted by NaN, in the Age and Cabin columns. <br>
The names of the columns SibSp, Parch, Cabin and Embarked do not bring out their utility. So, the documentation of the dataset needs to be checked. <br> The Fare column should be examined for inconsistencies - passenger with ID 272 did not have to pay any fare.

## Separate the target variable from the remaining columns

In [152]:
target = df_titanic['Survived']

In [153]:
df_titanic.drop(columns=['Survived'], inplace=True) #inplace=True means that df will change

In [154]:
df_titanic.head(8)

PassengerId  Pclass                                               Name  \
0            1       3                            Braund, Mr. Owen Harris   
1            2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3       3                             Heikkinen, Miss. Laina   
3            4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            5       3                           Allen, Mr. William Henry   
5            6       3                                   Moran, Mr. James   
6            7       1                            McCarthy, Mr. Timothy J   
7            8       3                     Palsson, Master. Gosta Leonard   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S  
5    male   NaN      0      0            330877   8.4583   NaN        Q  
6    male  54.0      0      0             17463  51.8625   E46        S  
7    male   2.0      3      1            349909  21.0750   NaN        S

## Finding the utility of the columns, if needed

* **SibSp**: Indicates number of siblings/spouses aboard; used to derive **family size**, which affects survival probability.
* **Parch**: Indicates number of parents/children aboard; combined with SibSp to model **family presence vs solo travel**.
* **Cabin**: Proxy for **passenger location in ship and socio-economic status**; all passengers did not have a Cabin.
* **Embarked**: Port of embarkation; used as a **categorical socio-economic and travel-pattern feature**. C = Cherbourg / Q = Queenstown / S = Southampton.

## Extracting high level information about the dataset

In [155]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Pclass       891 non-null    int64  
 2   Name         891 non-null    object 
 3   Sex          891 non-null    object 
 4   Age          714 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Cabin        204 non-null    object 
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 76.7+ KB


<i>Number of rows, number of columns, number of NULL values in every column, data type of every column</i>

## Check for duplicates

### Check row-wise

In [156]:
df_titanic.duplicated().sum()

0

In [157]:
# view the duplicate records, if any
df_titanic[df_titanic.duplicated()]

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

*There are no duplicate rows in the titanic dataset.*

### Check for specific column

In [158]:
df_titanic[df_titanic.duplicated(subset=['PassengerId'], keep=False)]

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

*Passenger id is analogous to primary key of a relation.*

In [159]:
df_titanic[df_titanic.duplicated(subset=['Ticket'], keep=False)] #i want to keep all instances of duplicated record in the output

PassengerId  Pclass                                               Name  \
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
7              8       3                     Palsson, Master. Gosta Leonard   
8              9       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9             10       2                Nasser, Mrs. Nicholas (Adele Achem)   
10            11       3                    Sandstrom, Miss. Marguerite Rut   
..           ...     ...                                                ...   
876          877       3                      Gustafsson, Mr. Alfred Ossian   
879          880       1      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)   
880          881       2       Shelley, Mrs. William (Imanita Parrish Hall)   
885          886       3               Rice, Mrs. William (Margaret Norton)   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
3    female  35.0      1      0      113803  53.1000  C123        S  
7      male   2.0      3      1      349909  21.0750   NaN        S  
8    female  27.0      0      2      347742  11.1333   NaN        S  
9    female  14.0      1      0      237736  30.0708   NaN        C  
10   female   4.0      1      1     PP 9549  16.7000    G6        S  
..      ...   ...    ...    ...         ...      ...   ...      ...  
876    male  20.0      0      0        7534   9.8458   NaN        S  
879  female  56.0      0      1       11767  83.1583   C50        C  
880  female  25.0      0      1      230433  26.0000   NaN        S  
885  female  39.0      0      5      382652  29.1250   NaN        Q  
888  female   NaN      1      2  W./C. 6607  23.4500   NaN        S  

[344 rows x 11 columns]

*It appears that multiple passengers were travelling on the same ticket. This is logically consistent because many were travelling with family - siblings, spouse, children and parents. This does not represent duplicate data.*

**keep=False displays all duplicate records instead of keeping only the first occurrence or only the last occurrence. This helps to determine the actual number of duplicates.**

### Check for multiple columns

In [160]:
df_titanic[df_titanic.duplicated(subset=["Name", "Age", "Sex", "Pclass"], keep=False)]

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

*Display those rows for which the Name, Age, Sex, Pclass attributes have the exact same values.*

**The Titanic dataset contains no exact duplicate records; apparent duplicates arise from shared tickets or family relationships and must not be removed.**

## Handling missing values

### Identify missing values

In [161]:
df_titanic.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

- Only 3 columns contain missing values out of which Embarked contains only 2 missing values. These two rows can be examined separately, if needed.
- Age contains 177 missing values. Reasons could be many. Age may not have been a mandatory information for adults, poorer passengers may not have had proper documentation, children's ages may have been estimated or omitted, etc.
- It is infeasible to drop 177 records corresponding to missing age values. Rather, age can be estimated or "imputed".
- Most passengers did not have access to a cabin. This explains the large number of missing values in the Cabin attribute.


### Imputing Age using median value

In [162]:
df_titanic['Age'] = df_titanic['Age'].fillna(df_titanic['Age'].median()) #using median age of passengers and imputing it in the missing values
#but mean cannot be used if there are outliers

In [163]:
df_titanic.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

*There are other ways to impute numerical columns containing missing values - conditional imputation, model-based imputation and probabilistic imputation.*

### Imputing Age using Sex and Pclass

In [164]:
df_titanic_2 = pd.read_csv(path_of_dataset)

In [165]:
df_titanic_2.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

#### Compute group wise median age

In [166]:
group_median_age = df_titanic_2.groupby(['Sex', 'Pclass'])['Age'].median()
#-groupby is used to put the data in groups
#grouping based of 2 attributes and out of the groups we are finding the age

In [167]:
print(group_median_age)

Sex     Pclass
female  1         35.0
        2         28.0
        3         21.5
male    1         40.0
        2         30.0
        3         25.0
Name: Age, dtype: float64


#### Defining an imputation function

In [168]:
def impute_age(row):
  if pd.isnull(row['Age']) == True:
    return group_median_age.loc[row['Sex'], row['Pclass']]
  else:
    return row['Age']

In [169]:
df_titanic_2['Age'] = df_titanic_2.apply(impute_age, axis=1)
#FUNCTION IS PASSED IN EACH ROW

In [170]:
df_titanic_2.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### Handling the Cabin column

*The Cabin column contains a very large number of missing values. It is a categorical attribute. From domain knowledge, it is found that the many passengers who were not travelling first class did not have a dedicated cabin. This means allocation of cabin is an indication of socio-economic status and having or not having a cabin may influence the chances of survival of a passenger. Thus, engineering a feature from the Cabin attribute may be a good option.*

In [171]:
df_titanic['CabinAllocated'] = df_titanic['Cabin'].notna().map({True:"YES", False:"NO"})
#ADDING A NEW COLUMN WHICH WILL INDICATE WHETHER CABIN HAS BEEN ALLOCATED OR NOT

In [172]:
df_titanic.head()

PassengerId  Pclass                                               Name  \
0            1       3                            Braund, Mr. Owen Harris   
1            2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3       3                             Heikkinen, Miss. Laina   
3            4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            5       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

  CabinAllocated  
0             NO  
1            YES  
2             NO  
3            YES  
4             NO

### Handling the Embarked column

In [173]:
# Finding those two rows where Embarked attribute contains NaN
df_titanic[df_titanic['Embarked'].isna()]

PassengerId  Pclass                                       Name     Sex  \
61            62       1                        Icard, Miss. Amelie  female   
829          830       1  Stone, Mrs. George Nelson (Martha Evelyn)  female   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked CabinAllocated  
61   38.0      0      0  113572  80.0   B28      NaN            YES  
829  62.0      0      0  113572  80.0   B28      NaN            YES

[Passenger information](https://www.encyclopedia-titanica.org/titanic-survivor/amelia-icard.html)

*Two passengers can be related by employer-employee relationship as well!*

In [174]:
df_titanic.loc[61, 'Embarked'] = 'S'
df_titanic.loc[829, 'Embarked'] = 'S'

In [175]:
df_titanic[df_titanic['Embarked'].isna()]

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked, CabinAllocated]
Index: []

In [176]:
df_titanic.isnull().sum()

PassengerId         0
Pclass              0
Name                0
Sex                 0
Age                 0
SibSp               0
Parch               0
Ticket              0
Fare                0
Cabin             687
Embarked            0
CabinAllocated      0
dtype: int64

*The Cabin column can be removed from the dataset.*

In [177]:
df_titanic.drop(columns=['Cabin'], inplace=True)

In [185]:
df_titanic.isnull().sum()

PassengerId       0
Pclass            0
Name              0
Sex               0
Age               0
SibSp             0
Parch             0
Ticket            0
Fare              0
Embarked          0
CabinAllocated    0
dtype: int64

*The dataset is free from missing values!*

## Checking and handling invalid values

In [179]:
df_titanic.shape

(891, 11)

In [180]:
df_titanic.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Embarked', 'CabinAllocated'],
      dtype='object')

In [181]:
df_titanic.dtypes

PassengerId         int64
Pclass              int64
Name               object
Sex                object
Age               float64
SibSp               int64
Parch               int64
Ticket             object
Fare              float64
Embarked           object
CabinAllocated     object
dtype: object

*The columns Age and Fare are numerical and can be checked for inconsistencies such as Age should be greater than 0 and Fare should also be greater than 0.*

In [182]:
df_titanic.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  891.000000  891.000000  891.000000
mean    446.000000    2.308642   29.361582    0.523008    0.381594   32.204208
std     257.353842    0.836071   13.019697    1.102743    0.806057   49.693429
min       1.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%     223.500000    2.000000   22.000000    0.000000    0.000000    7.910400
50%     446.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%     668.500000    3.000000   35.000000    1.000000    0.000000   31.000000
max     891.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- The minimum and maximum values of the Age attribute are logically consistent.
- The minimum Fare is 0 which does not agree with common knowledge. This requires further investigation.

In [183]:
df_titanic[df_titanic['Fare'] <= 0]

PassengerId  Pclass                              Name   Sex   Age  SibSp  \
179          180       3               Leonard, Mr. Lionel  male  36.0      0   
263          264       1             Harrison, Mr. William  male  40.0      0   
271          272       3      Tornquist, Mr. William Henry  male  25.0      0   
277          278       2       Parkes, Mr. Francis "Frank"  male  28.0      0   
302          303       3   Johnson, Mr. William Cahoone Jr  male  19.0      0   
413          414       2    Cunningham, Mr. Alfred Fleming  male  28.0      0   
466          467       2             Campbell, Mr. William  male  28.0      0   
481          482       2  Frost, Mr. Anthony Wood "Archie"  male  28.0      0   
597          598       3               Johnson, Mr. Alfred  male  49.0      0   
633          634       1     Parr, Mr. William Henry Marsh  male  28.0      0   
674          675       2        Watson, Mr. Ennis Hastings  male  28.0      0   
732          733       2              Knight, Mr. Robert J  male  28.0      0   
806          807       1            Andrews, Mr. Thomas Jr  male  39.0      0   
815          816       1                  Fry, Mr. Richard  male  28.0      0   
822          823       1   Reuchlin, Jonkheer. John George  male  38.0      0   

     Parch  Ticket  Fare Embarked CabinAllocated  
179      0    LINE   0.0        S             NO  
263      0  112059   0.0        S            YES  
271      0    LINE   0.0        S             NO  
277      0  239853   0.0        S             NO  
302      0    LINE   0.0        S             NO  
413      0  239853   0.0        S             NO  
466      0  239853   0.0        S             NO  
481      0  239854   0.0        S             NO  
597      0    LINE   0.0        S             NO  
633      0  112052   0.0        S             NO  
674      0  239856   0.0        S             NO  
732      0  239855   0.0        S             NO  
806      0  112050   0.0        S            YES  
815      0  112058   0.0        S            YES  
822      0   19972   0.0        S             NO

In [184]:
len(df_titanic[df_titanic['Fare'] <= 0])

15

Important observations:
- There are 15 male passengers who did not pay any fare.
- They appear to be solo travellers but may be accompanying an employer like we saw in case of Ms. Icard.
- It may be the case these males are workers on the titanic and therfore do not need a ticket but this claim does not hold in all cases because there exists two contradictions: all these passengers have a ticket and few of them even possess first class tickets (highly unlikely that workers will be travelling first class!). But it may be the case that few travellers were given complementary tickets by the shipping company (this may justify first class travel!). Also, all passengers may have been given tickets regardless of them paying fare or not.
- Passengers are also sharing tickets.

*Conclusion: It is best to not drop the rows with zero Fare because it represents vital information regarding the types of passengers who travelled on the Titanic.*

## Handling outliers

*Outliers refer to extremely high or extremely low numerical values found in attributes of the dataset. Statistically, outliers may be present but these may often represent actual data. In the Titanic dataset, very high fares may represent passengers who actually bought a very expensive ticket with a lot facilites. Thus, outliers should be dropped keeping consistency with the domain knowledge of the problem statement.*

### Detect outliers

- Outliers can be detected using the formula for Interquartile range (IQR)
- IQR = Q3 - Q1
- Q1 (1st quartile) = 25th percentile = 25% of data lie below this value
- Q3 (3rd quartile) = 75th percentile = 75% of data lie below this value
- IQR measures the spread of the middle 50% of the data
- Lower bound = Q1 - 1.5 * IQR
- Upper bound = Q3 + 1.5 * IQR
- Any value < Lower bound → lower outlier
- Any value > Upper bound → upper outlier
- These points lie far outside the central bulk of the data.

In [ ]:
Q1 = df_titanic['Fare'].quantile(0.25)
Q3 = df_titanic['Fare'].quantile(0.75)
print(Q1, Q3)

7.9104 31.0


In [ ]:
IQR = Q3 - Q1
print(IQR)

23.0896


In [ ]:
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
print(f'Lower bound is {lower_bound} and upper bound is {upper_bound}')

Lower bound is -26.724 and upper bound is 65.6344


In [ ]:
fare_outliers = df_titanic[(df_titanic['Fare'] < lower_bound) | (df_titanic['Fare'] > upper_bound)]

In [ ]:
fare_outliers.sample(10)

PassengerId  Survived  Pclass  \
325          326         1       1   
665          666         0       2   
779          780         1       1   
763          764         1       1   
230          231         1       1   
373          374         0       1   
679          680         1       1   
375          376         1       1   
380          381         1       1   
310          311         1       1   

                                                  Name     Sex   Age  SibSp  \
325                           Young, Miss. Marie Grice  female  36.0      0   
665                                 Hickman, Mr. Lewis    male  32.0      2   
779  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...  female  43.0      0   
763          Carter, Mrs. William Ernest (Lucile Polk)  female  36.0      1   
230       Harris, Mrs. Henry Birkhardt (Irene Wallach)  female  35.0      1   
373                                Ringhini, Mr. Sante    male  22.0      0   
679                 Cardeza, Mr. Thomas Drake Martinez    male  36.0      0   
375              Meyer, Mrs. Edgar Joseph (Leila Saks)  female  28.0      1   
380                              Bidois, Miss. Rosalie  female  42.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   

     Parch        Ticket      Fare Embarked CabinAllocated  
325      0      PC 17760  135.6333        C            YES  
665      0  S.O.C. 14879   73.5000        S             NO  
779      1         24160  211.3375        S            YES  
763      2        113760  120.0000        S            YES  
230      0         36973   83.4750        S            YES  
373      0      PC 17760  135.6333        C             NO  
679      1      PC 17755  512.3292        C            YES  
375      0      PC 17604   82.1708        C             NO  
380      0      PC 17757  227.5250        C             NO  
310      0         11767   83.1583        C            YES

*The Fare variable is highly right skewed. First class passengers paid much higher fares. Many tickets were shared, inflating per passenger fare. IQR is aggressive for skewed data. Therefore, IQR flags many high but valid fares as outliers. These values should be transformed or capped rather than removed.*

### Cap outliers

In [ ]:
df_titanic['Fare'] = df_titanic['Fare'].clip(lower_bound, upper_bound)

In [ ]:
df_titanic.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.361582    0.523008   
std     257.353842    0.486592    0.836071   13.019697    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   24.046813  
std      0.806057   20.481625  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000   65.634400

## Encoding categorical variables

In [ ]:
df_titanic.sample(5)

PassengerId  Survived  Pclass                               Name     Sex  \
115          116         0       3              Pekoniemi, Mr. Edvard    male   
341          342         1       1     Fortune, Miss. Alice Elizabeth  female   
889          890         1       1              Behr, Mr. Karl Howell    male   
446          447         1       2  Mellinger, Miss. Madeleine Violet  female   
593          594         0       3                 Bourke, Miss. Mary  female   

      Age  SibSp  Parch             Ticket     Fare Embarked CabinAllocated  
115  21.0      0      0  STON/O 2. 3101294   7.9250        S             NO  
341  24.0      3      2              19950  65.6344        S            YES  
889  26.0      0      0             111369  30.0000        C            YES  
446  13.0      0      1             250644  19.5000        S             NO  
593  28.0      0      2             364848   7.7500        Q             NO

Key takeaways
- Categorical variables take fixed number of predefined values.
- Categorical variables are of two types: nominal and ordinal.
- The values in these variables have to be converted to numerical representations depending on the type of categorical variable because data modelling requires numbers and cannot work with strings.
- The order does not matter in nominal variables.
- The order matters in ordinal variables.
- There are 5 categorical variables: Pclass, Sex, Ticket, Embarked and CabinAllocated.
- Pclass is ordinal and already in numerical form.
- Sex is binary nominal and needs to be label encoded.
- Ticket is nominal with high cardinality but is not considered a very important variable. It will be dropped.
- Embarked is multi class nominal and needs to be one-hot encoded.
- CabinAllocated is binary nominal and needs to be label encoded.
- Binary nominal variables can be one-hot encoded but this is not required as label encoding and one-hot encoding represent equivalent information in this case.
- Multi class nominal variables should not be label encoded because it introduces an ordering amongst the categories and may mislead the algorithm used for data modelling.

### Label encoding the variable Sex

In [ ]:
df_titanic['Sex'] = df_titanic['Sex'].map({'male':0, 'female':1})

In [ ]:
df_titanic.sample(5)

PassengerId  Survived  Pclass  \
545          546         0       1   
576          577         1       2   
361          362         0       2   
802          803         1       1   
608          609         1       2   

                                                  Name  Sex   Age  SibSp  \
545                       Nicholson, Mr. Arthur Ernest    0  64.0      0   
576                               Garside, Miss. Ethel    1  34.0      0   
361                          del Carlo, Mr. Sebastiano    0  29.0      1   
802                Carter, Master. William Thornton II    0  11.0      1   
608  Laroche, Mrs. Joseph (Juliette Marie Louise La...    1  22.0      1   

     Parch         Ticket     Fare Embarked CabinAllocated  
545      0            693  26.0000        S             NO  
576      0         243880  13.0000        S             NO  
361      0  SC/PARIS 2167  27.7208        C             NO  
802      2         113760  65.6344        S            YES  
608      2  SC/Paris 2123  41.5792        C             NO

### One-hot encoding the variable Embarked

In [ ]:
df_titanic = pd.get_dummies(df_titanic, columns=['Embarked'], drop_first=True)

In [ ]:
df_titanic.sample(5)

PassengerId  Survived  Pclass  \
307          308         1       1   
719          720         0       3   
246          247         0       3   
437          438         1       2   
300          301         1       3   

                                                  Name  Sex   Age  SibSp  \
307  Penasco y Castellana, Mrs. Victor de Satode (M...    1  17.0      1   
719                       Johnson, Mr. Malkolm Joackim    0  33.0      0   
246              Lindahl, Miss. Agda Thorilda Viktoria    1  25.0      0   
437              Richards, Mrs. Sidney (Emily Hocking)    1  24.0      2   
300           Kelly, Miss. Anna Katherine "Annie Kate"    1  28.0      0   

     Parch    Ticket     Fare CabinAllocated  Embarked_Q  Embarked_S  
307      0  PC 17758  65.6344            YES       False       False  
719      0    347062   7.7750             NO       False        True  
246      0    347071   7.7750             NO       False        True  
437      3     29106  18.7500             NO       False        True  
300      0      9234   7.7500             NO        True       False

Key takeaways
- When you one-hot encode a categorical variable with k categories, you get k dummy columns that are linearly dependent.
- Embarked_C = 1 − Embarked_Q − Embarked_S
- One column is perfectly predictable from the others.
- The dummy variable trap occurs when all the one-hot encoded variables are included in the dataset.
- It leads to the problem of multicollinearity.
- Multicollinearity: one input column can be predicted from other input columns, so the model gets redundant information and becomes unstable.
- drop_first=True (solution)
- Drops one category (here Embarked_C) and uses it as the reference category.
- Embarked_Q = 1 → passenger embarked at Q
- Embarked_S = 1 → passenger embarked at S
- Both 0 → passenger embarked at C
- One-hot encoding may produce (pandas does) Boolean values, where True represents 1 and False represents 0; True/False is just pandas’ way of writing 1/0; no need to convert to numerical values.

### Label encoding the variable CabinAllocated

In [ ]:
df_titanic['CabinAllocated'] = df_titanic['CabinAllocated'].map({'NO':0, 'YES':1})

In [ ]:
df_titanic.sample(5)

PassengerId  Survived  Pclass                                     Name  \
326          327         0       3                Nysveen, Mr. Johan Hansen   
63            64         0       3                    Skoog, Master. Harald   
333          334         0       3          Vander Planke, Mr. Leo Edmondus   
448          449         1       3           Baclini, Miss. Marie Catherine   
775          776         0       3  Myhrman, Mr. Pehr Fabian Oliver Malkolm   

     Sex   Age  SibSp  Parch  Ticket     Fare  CabinAllocated  Embarked_Q  \
326    0  61.0      0      0  345364   6.2375               0       False   
63     0   4.0      3      2  347088  27.9000               0       False   
333    0  16.0      2      0  345764  18.0000               0       False   
448    1   5.0      2      1    2666  19.2583               0       False   
775    0  18.0      0      0  347078   7.7500               0       False   

     Embarked_S  
326        True  
63         True  
333        True  
448       False  
775        True

## Removing irrelevant columns

In [ ]:
df_titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'CabinAllocated', 'Embarked_Q',
       'Embarked_S'],
      dtype='object')

In [ ]:
df_titanic_backup = df_titanic.copy()

In [ ]:
df_titanic.drop(columns=['PassengerId', 'Ticket', 'Name'], inplace=True)

In [ ]:
df_titanic.sample(2)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  CabinAllocated  \
691         1       3    1   4.0      0      1  13.4167               0   
463         0       2    0  48.0      0      0  13.0000               0   

     Embarked_Q  Embarked_S  
691       False       False  
463       False        True

## Bining and Smoothing

- These techniques are applied only to numerical columns.
- Bining groups continuous values into intervals.
- Smoothing replaces values in each bin to reduce noise.

### Equal width bining (Age)

In [ ]:
df_titanic['Age_bin'] = pd.cut(df_titanic['Age'],
                               bins=[0, 12, 18, 35, 60, 100],
                               labels=['Child', 'Teen', 'YoungAdult', 'Adult', 'Senior'])

*This reduces noise in Age and converts it into meaningful groups.*

**pd.cut creates right-inclusive intervals: (a, b] = greater than a and less than or equal to b.**

### Bin-median smoothing (Age)

In [ ]:
df_titanic['Age_smoothed'] = df_titanic.groupby('Age_bin', observed=True)['Age'].transform('median')

*This replaces each Age value with the median Age of its bin.*

In [ ]:
df_titanic.sample(2)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  CabinAllocated  \
65         1       3    0  28.0      1      1  15.2458               0   
54         0       1    0  65.0      0      1  61.9792               1   

    Embarked_Q  Embarked_S     Age_bin  Age_smoothed  
65       False       False  YoungAdult          28.0  
54       False       False      Senior          64.5

### Equal frequency bining using quantiles (Fare)

In [ ]:
df_titanic['Fare_bin'] = pd.qcut(df_titanic['Fare'], q=4, labels=['Low', 'Mid', 'High', 'VeryHigh'])

### Bin-mean smoothing (Fare)

In [ ]:
df_titanic['Fare_smoothed'] = df_titanic.groupby('Fare_bin', observed=True)['Fare'].transform('mean')

In [ ]:
df_titanic.sample(2)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  CabinAllocated  \
741         0       1    0  36.0      1      0  65.6344               1   
811         0       3    0  39.0      0      0  24.1500               0   

     Embarked_Q  Embarked_S Age_bin  Age_smoothed  Fare_bin  Fare_smoothed  
741       False        True   Adult          44.0  VeryHigh      55.938059  
811       False        True   Adult          44.0      High      23.028677

**observed=True** implies that only those bins are included where there are values.

In [ ]:
df_titanic['Age_bin'].value_counts()

Age_bin
YoungAdult    535
Adult         195
Teen           70
Child          69
Senior         22
Name: count, dtype: int64

In [ ]:
df_titanic['Fare_bin'].value_counts()

Fare_bin
Mid         224
Low         223
High        222
VeryHigh    222
Name: count, dtype: int64

*What to include and what to exclude?*